### PR TITLE
[V1] Logit processors for rejection sampler

### PR DESCRIPTION
**Pull request by southfreebird**
Hi team,
Thank you for your amazing work!
This PR was prepared by the Nebius team.

This is a draft of the PR, so your suggestions are welcome! I would be happy to improve it or add anything I missed.

## Purpose
This PR adds logit processors to the rejection sampler, which allows the use of bad_words, penalties, and allowed_token_ids with speculative decoding. Previously, these parameters were silently ignored. The problem was described in issue: https://github.com/vllm-project/vllm/issues/18902

## Test Plan
You can test this functionality using the following pytest command:
```
pytest tests/v1/sample/test_rejection_sampler.py -k "test_frequency_penalties or test_bad_words or test_allowed_token_ids"
```
Or use the script provided in the issue. 

## Test Result

Server command:
```
vllm serve meta-llama/Llama-3.1-8B-Instruct --speculative-config '{"method": "eagle3", "model": "yuhuili/EAGLE3-LLaMA3.1-Instruct-8B", "num_speculative_tokens": 4}' --max-model-len 2048 --no-enable-prefix-caching
```

Client command:
```
vllm bench serve --model meta-llama/Llama-3.1-8B-Instruct --dataset-name sharegpt --num-prompts 200 --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json --seed 0
```

Baseline:
```
============ Serving Benchmark Result ============
Successful requests:                     200       
Benchmark duration (s):                  10.76     
Total input tokens:                      42659     
Total generated tokens:                  43382     
Request throughput (req/s):              18.59     
Output token throughput (tok/s):         4032.80   
Peak output token throughput (tok/s):    4726.00   
Peak concurrent requests:                200.00    
Total Token throughput (tok/s):          7998.39   
---------------Time to First Token----------------
Mean TTFT (ms):                          887.85    
Median TTFT (ms):                        854.29    
P99 TTFT (ms):                           1306.54   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          28.76     
Median TPOT (ms):                        18.41     
P99 TPOT (ms):                           159.41    
---------------Inter-token Latency----------------
Mean ITL (ms):                           26.70     
Median ITL (ms):                         24.23     
P99 ITL (ms):                            207.50    
==================================================
```

Measurements with a frequency-penalty enabled:
```
============ Serving Benchmark Result ============
Successful requests:                     200       
Benchmark duration (s):                  11.50     
Total input tokens:                      42659     
Total generated tokens:                  43052     
Request throughput (req/s):              17.39     
Output token throughput (tok/s):         3743.33   
Peak output token throughput (tok/s):    4310.00   
Peak concurrent requests:                200.00    
Total Token throughput (tok/s):          7452.49   
---------------Time to First Token----------------
Mean TTFT (ms):                          902.73    
Median TTFT (ms):                        874.15    
P99 TTFT (ms):                           1341.88   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          31.91     
Median TPOT (ms):                        21.44     
P99 TPOT (ms):                           161.59    
---------------Inter-token Latency----------------
Mean ITL (ms):                           27.22     
Median ITL (ms):                         23.93     
P99 ITL (ms):                            207.16    
==================================================
```

### CODE REVIEW

File: vllm/v1/sample/rejection_sampler.py [hunk 1]
```diff
@@ -83,6 +86,14 @@ def forward(
                 A tensor containing the final output token IDs.
         '''
         assert metadata.max_spec_len <= MAX_SPEC_LEN
+
+        # Use float32 for the target_logits.
+        target_logits = target_logits.to(torch.float32)
```
Comment (southfreebird) [inline vllm/v1/sample/rejection_sampler.py]: I'm not sure about this change.
I added it because of the new logic in https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/utils.py#L56-L57.

It requires logits and repetition_penalties to have the same dtype. Without casting logits to float32, we get logits in bfloat16 and repetition_penalties in float32. To fix that, we could cast repetition_penalties to bfloat16, but we would lose some accuracy here. Casting logits to float32 makes it consistent with the sampler logic.

**Comment by southfreebird** [comment]
Hi team!
Any chance this PR to be reviewed?

**Comment by ilyal-cerebras** [comment]
Do you still observe benefits from spec decode when penalties are applied?
There is an opinion that when penalties are applied, draft and target models become dis-synchronized and a number of confirmed candidates is reduced.

**Comment by southfreebird** [comment]
> Do you still observe benefits from spec decode when penalties are applied?
There is an opinion that when penalties are applied, draft and target models become dis-synchronized and a number of confirmed candidates is reduced.

Yeah, you are right. This PR just applies penalties to the bonus token when penalties are applied.

**Review by benchislett — COMMENTED** [review]
Why do you need to track `last_spec_token_ids` instead of using `scheduler_output.scheduled_spec_decode_tokens`?

**Comment by southfreebird** [comment]
Hey @benchislett
My main idea was to keep the interface of Samplers consistent and to pass all arguments required for sampling through sampling_metadata. Do you suggest passing them directly to the samplers as  scheduler_output.scheduled_spec_decode_tokens instead?

File: vllm/v1/sample/rejection_sampler.py (hunk 1 above)
Comment (benchislett) [reply vllm/v1/sample/rejection_sampler.py]: This seems reasonable, but I'm not too sure about the potential consequences. FWIW, I don't think we would lose too much accuracy by casting repetition_penalties to BF16 since they are usually very small values (often < 2)

**Review by benchislett — APPROVED** [review]
I see now. It is somewhat confusing to me that `self.input_batch.last_spec_token_ids` is named this way, as it looks like it's just the current spec token ids that need to be verified. If possible I would suggest renaming to `spec_token_ids` to clarify if possible.

Also, please include a benchmark with performance before/after this change. It is unclear to me how much (if any) slowdown will be seen when we add this feature.

Otherwise, LGTM. I think someone else should take a look at the logits-casting to make sure it won't have consequences in other areas.

**Comment by benchislett** [comment]
@southfreebird in your benchmark, please ensure that some logit processors are enabled. Otherwise I worry this might not capture the performance change.

**Comment by afeldman-nm** [comment]
JFYI @southfreebird the approach for supporting built-in logits processors in vLLM v1 is changing, as described here:

https://docs.vllm.ai/en/latest/design/logits_processors.html

So in the long term, bad words, penalties and allowed token ids should be updated to be subclasses of `LogitsProcessor` as described in #25957 

However, #25957 depends on #25389 (revision to the programming model for defining subclasses of `LogitsProcessor`) which is still WIP.

So for the purposes of this PR I think the approach taken (augment the existing bad words/penalties/allowed token ids support to be compatible with spec decoding) is fine, since the logits processor programming model will be revised soon.

**Comment by simon-mo** [comment]
@22quinn @houseroad please help review logits/sampler changes

**Comment by njhill** [comment]
I'll review this today/tomorrow

**Comment by southfreebird** [comment]
Hey @njhill, just bumping this PR for visibility when you get a chance to look at it!

File: vllm/v1/sample/rejection_sampler.py [hunk 2]
```diff
@@ -131,6 +141,123 @@ def parse_output(
...
+            )
+        # Calculate indices of target logits
+        if sampling_metadata.allowed_token_ids_mask is not None \
+            or not sampling_metadata.no_penalties:
+            num_requests = len(sampling_metadata.output_token_ids)
+            num_draft_tokens = torch.tensor(metadata.num_draft_tokens,
+                                            device="cpu")
+            original_indices = torch.arange(num_requests, device="cpu")
+            repeat_indices_cpu = torch.repeat_interleave(
+                original_indices, repeats=num_draft_tokens)
```
Comment (njhill) [inline vllm/v1/sample/rejection_sampler.py]: nit: slightly simpler
```suggestion
            repeat_indices_cpu = original_indices.repeat_interleave(num_draft_tokens)
```

File: vllm/v1/sample/rejection_sampler.py [hunk 3]
```diff
@@ -131,6 +141,123 @@ def parse_output(
...
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        repeat_indices: torch.Tensor,
+    ) -> torch.Tensor:
+        if sampling_metadata.allowed_token_ids_mask is not None:
+            allowed_token_ids_mask = sampling_metadata.allowed_token_ids_mask[
+                repeat_indices]
+            logits.masked_fill_(allowed_token_ids_mask, float("-inf"))
+
+        return logits
```
Comment (njhill) [inline vllm/v1/sample/rejection_sampler.py]: probably simpler to just inline these two in `apply_logits_processors` ... same for the ones in sampler.py

File: vllm/v1/sample/rejection_sampler.py [hunk 4]
```diff
@@ -131,6 +141,123 @@ def parse_output(
...
+            repetition_penalties,
+            output_token_ids,
+        )
+        return logits
+
+    def _combine_outputs_with_spec_tokens(
+        self,
+        output_token_ids: list[list[int]],
+        spec_token_ids: Optional[list[list[int]]] = None,
+    ):
```
Comment (njhill) [inline vllm/v1/sample/rejection_sampler.py]:
```suggestion
    ) -> list[list[int]]]:
```

File: vllm/v1/sample/sampler.py [hunk 5]
```diff
@@ -244,10 +236,56 @@ def gather_logprobs(
 
         return LogprobsTensors(indices, logprobs, token_ranks)
 
+    def _combine_outputs_with_spec_tokens(
+        self,
+        output_token_ids: list[list[int]],
+        spec_token_ids: Optional[list[list[int]]] = None,
+    ):
```
Comment (njhill) [inline vllm/v1/sample/sampler.py]:
```suggestion
    ) -> list[list[int]]:
```

File: vllm/v1/sample/sampler.py [hunk 6]
```diff
@@ -244,10 +236,56 @@ def gather_logprobs(
...
+
+        return [[*out, *spec] if spec else out
+                for out, spec in zip(output_token_ids, spec_token_ids)]
+
+    def apply_logits_processors(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        predict_bonus_token: bool,
+    ):
```
Comment (njhill) [inline vllm/v1/sample/sampler.py]:
```suggestion
    ) -> torch.Tensor:
```

File: vllm/v1/sample/rejection_sampler.py [hunk 7]
```diff
@@ -131,6 +141,123 @@ def parse_output(
         ]
         return outputs
 
+    def apply_logits_processors(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        metadata: SpecDecodeMetadata,
+    ):
```
Comment (njhill) [inline vllm/v1/sample/rejection_sampler.py]:
```suggestion
    ) -> torch.Tensor:
```

File: vllm/v1/sample/rejection_sampler.py [hunk 8]
```diff
@@ -131,6 +141,123 @@ def parse_output(
...
+        else:
+            # Not use them for other cases
+            repeat_indices = None
+
+        logits = self.apply_penalties(logits, sampling_metadata, metadata,
+                                      repeat_indices, output_token_ids)
+        logits = self.apply_bad_words(logits, sampling_metadata, metadata,
+                                      output_token_ids)
+        logits = self.apply_allowed_token_ids(logits, sampling_metadata,
+                                              repeat_indices)
```
Comment (njhill) [inline vllm/v1/sample/rejection_sampler.py]: Can simplify here I think
```suggestion
            logits = self.apply_penalties(logits, sampling_metadata, metadata,
                                  repeat_indices, output_token_ids)
            logits = self.apply_allowed_token_ids(logits, sampling_metadata,
                                                  repeat_indices)

        logits = self.apply_bad_words(logits, sampling_metadata, metadata,
                                      output_token_ids)
```

File: vllm/v1/sample/rejection_sampler.py [hunk 9]
```diff
@@ -131,6 +141,123 @@ def parse_output(
...
+
+        return logits
+
+    def apply_penalties(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        metadata: SpecDecodeMetadata,
+        repeat_indices: Optional[torch.Tensor],
```
Comment (njhill) [inline vllm/v1/sample/rejection_sampler.py]: per other change
```suggestion
        repeat_indices: torch.Tensor,
```

**Review by njhill — COMMENTED** [review]
Thanks @southfreebird and sorry for taking so long to get to this.

I think we can do things better performance-wise but this part of the code is pending a refactor anyhow, related to the logit processor rework that @afeldman-nm mentioned, model runner refactor, and async scheduling compatibility. 

In particular we'll want to look at how this could also work with generic/custom logit processors.

But we can get this merged first and address those as follow-on work.

As well as the inline comments, it would be good to add a startup check for spec decode + custom logits processors (if there is not one already) since it will still be silently "incorrect" in that case otherwise.

**Comment by southfreebird** [comment]
Hey @njhill 
Thank you for your review!

I have added an error and logging for the scenario when custom logit processors are enabled. Please, check if it looks good for you. 

And you are right about the new interface @afeldman-nm mentioned, it should definitely be unified (or extended by new methods) to be suitable for use in a rejection sampler. I'd love to help you with that if you don't mind.

File: vllm/v1/sample/logits_processor/__init__.py [hunk 10]
```diff
@@ -185,6 +191,17 @@ def build_logitsprocs(
             " do not support logits processors."
         )
         return LogitsProcessors()
+
+    # Check if speculative decoding is enabled.
+    if vllm_config.speculative_config:
+        if custom_logitsprocs:
+            raise ValueError(STR_SPEC_DEC_REJECTS_LOGITSPROCS)
+        logger.debug(
```
Comment (njhill) [inline vllm/v1/sample/logits_processor/__init__.py]: I just realized that this means the built-in logits processors also won't work which is a bit unfortunate since from the outside these are no different to other sampling params (penalties etc.)

The plan was/is to port most of the sampling param impls to be LPs once we've stabilized that interface, we had just done this with a couple of them initially in preparation for that. So there will still be some "silent failure" if these parameters are used (min_p, logit_bias, min_tokens), but I guess it's no worse than the current state so probably still ok to get this in and address as part of the larger refactoring underway.

File: vllm/v1/sample/logits_processor/__init__.py [hunk 11]
```diff
@@ -185,6 +191,17 @@ def build_logitsprocs(
...
+            "Skipping logits processor loading because "
+            "they are not supported when speculative decoding is enabled."
```
Comment (njhill) [inline vllm/v1/sample/logits_processor/__init__.py]: actually given the above we may want to change this to a warning and change the message
```suggestion
        logger.warning(
            "min_p, logit_bias, and min_tokens parameters won't currently work "
            "with speculative decoding enabled."
```

File: vllm/v1/sample/logits_processor/__init__.py (hunk 10 above)
Comment (southfreebird) [reply vllm/v1/sample/logits_processor/__init__.py]: Yes, you are right here. We definitely need to find a way to apply built-in logit processors to the rejection sampler. I'll suggest a solution here this week.

Comment (njhill) [reply vllm/v1/sample/logits_processor/__init__.py]: Thanks! keep in mind it may look a bit different given the other in-flight/planned rework. For example I expect we'll remove various python lists from the input batch and instead use numpy arrays.

**Review by njhill — APPROVED** [review]
Thanks again @southfreebird

**Comment by southfreebird** [comment]
Sorry, everyone, for mentioning. 
I just tried to fix the DCO check. 
Reverted everything back (except a hundred mentions and new reviewers)
